src/multi.c

```diff
@@ -423,8 +423,20 @@ void watchCommand(client *c) {
         addReplyError(c,"WATCH inside MULTI is not allowed");
         return;
     }
-    for (j = 1; j < c->argc; j++)
+    for (j = 1; j < c->argc; j++) {
+        /* check and delete the expired keys, in case we watch the expired keys,
+         * if the expired keys get deleted before EXEC, the transaction would fail,
+         * but the deletion should not affect the transaction.
+         *
+         * and considering stale data would not be deleted if clients are paused,
+         * so the WATCH command is now with may-replicate flag, means WATCH command
+         * would be paused when redis is write paused mode.
+         *
+         * TODO: since replicas cannot delete expired keys by themselves, 
+         * we need more work in future to handle the corner case. */
+        expireIfNeeded(c->db,c->argv[j]);
```

Comment:
If we fix the TODO for the replicas, we don't need the may-replicate flag, right?

@oranagra wrote:

> 1. rename `expireIsNeeded` which is misleading to `IsKeyExpired` (or `isKeyExpiredAndMaybeDeleteIt` :smile:)
> 2. WATCH can call that function on the keys, and record the result (one bit) somewhere.
> 3. when the key is deleted either by lazy-expire, active-expire, or the expiration check in EXEC, if that bit is set, we don't flag the transaction as dirty.

(1) The name is indeed misleading. It doesn't *expire* keys, so indeed `isKeyExpiredAndMaybeDeleteIt` is a better name. :) The comment above the function does explain that it doesn't always delete an expired key though. I don't mind renaming it but I wouldn't require it either.

(2-3) This would work, but where do we store this bit? It would need to be in the client struct so it doesn't affect other clients' watches.

A different approach is to record the timestamp of the WATCH somewhere, for example replacing the list `client->watched_keys` with a dict (key => time of watching). Then EXEC can see if the key was expired already before WATCH. Then we don't need to do anything special in WATCH or in replicas.

PS. I'll be hiking for the next two weeks so I won't be able to follow up on this until August.

Reply:
finding a place for one bit in a hackish way is always easy (there are at least 2 wasted bits in each pointer).
i also though about storing the time, but that'll indeed require more memory, and in fact all we need is just one bit.

Reply:
Use a bit hack if you want :-) but storing the time doesn't need WATCH to have the may-replicate flag and WATCH doesn't even need to call expireIfNeeded. It's just an idea. You can choose.

Reply:
that's true for my bit representing the logical state of the key on watch too.. it no longer cares if expireIfNeeded deletes the key or not, so no need for may-replicate and no issues with replicas.
both achieve the same result, one may require a bit more memory, but maybe we could later find more uses for the time than just the state (bit).